log4j-sdk/src/main/java/org/apache/logging/log4j/sdk/logger/AbstractLogger.java

```diff
@@ -0,0 +1,3935 @@
+    // This is within the 35 byte MaxInlineSize threshold. Modify with care!
+    private void logMessageSafely(
+            final String fqcn,
+            final @Nullable StackTraceElement location,
+            final Level level,
+            final @Nullable Marker marker,
+            final @Nullable Message message,
+            final @Nullable Throwable throwable) {
+        try {
+            logMessageTrackRecursion(fqcn, location, level, marker, message, throwable);
+        } finally {
+            // LOG4J2-1583 prevent scrambled logs when logging calls are nested (logging in toString())
+            recycle(message);
+        }
+    }
+
+    @PerformanceSensitive
+    // NOTE: This is a hot method. Current implementation compiles to 33 bytes of byte code.
+    // This is within the 35 byte MaxInlineSize threshold. Modify with care!
+    private void logMessageTrackRecursion(
+            final String fqcn,
+            final @Nullable StackTraceElement location,
+            final Level level,
+            final @Nullable Marker marker,
+            final @Nullable Message message,
+            final @Nullable Throwable throwable) {
+        try {
+            incrementRecursionDepth(); // LOG4J2-1518, LOG4J2-2031
+            tryLogMessage(fqcn, location, level, marker, message, throwable);
+        } finally {
+            decrementRecursionDepth();
+        }
+    }
+
+    @PerformanceSensitive
+    // NOTE: This is a hot method. Current implementation compiles to 15 bytes of byte code.
+    // This is within the 35 byte MaxInlineSize threshold. Modify with care!
+    private @Nullable StackTraceElement getLocation(final String fqcn, final @Nullable StackTraceElement location) {
+        return location != null ? location : requiresLocation() ? StackLocatorUtil.calcLocation(fqcn) : null;
+    }
+
+    @PerformanceSensitive
+    // NOTE: This is a hot method. Current implementation compiles to 33 bytes of byte code.
+    // This is within the 35 byte MaxInlineSize threshold. Modify with care!
+    private void recycle(final @Nullable Message message) {
+        messageFactory.recycle(message);
+    }
+
+    @PerformanceSensitive
+    // NOTE: This is a hot method. Current implementation compiles to 26 bytes of byte code.
+    // This is within the 35 byte MaxInlineSize threshold. Modify with care!
+    private void tryLogMessage(
+            final String fqcn,
+            final @Nullable StackTraceElement location,
+            final Level level,
+            final @Nullable Marker marker,
+            final @Nullable Message message,
+            final @Nullable Throwable throwable) {
+        try {
+            doLog(fqcn, getLocation(fqcn, location), level, marker, message, throwable);
+        } catch (final Throwable t) {
+            // LOG4J2-1990 Log4j2 suppresses all exceptions that occur once application called the logger
+            handleLogMessageException(t, fqcn, message);
+        }
+    }
```

Comment:
This call to `getLocation` is IMHO precocious:

- if `requiresLocation()` returns true, it computes the location of each log message,
- however `requiresLocation()` does not evaluate the filters that a user configured down the line (in `LoggerConfig` or `Appender` for instance), so we could compute location that is not logged.

A better solution would be to move the call further down the logging pipeline before a thread jump or after an appender's filter has been evaluated.

The bonus of such a change would be that if I remove the `getLocation(fqcn, location` call I can introduce a:

```java
private @Nullable Throwable getThrowable(message, throwable) {
    return throwable != null ? throwable : message != null : message.getThrowable() : null;
}
```
method call and still maintain the 35 bytes bytecode limit. This will simplify the logic of many methods below.

Reply:
1. Does moving `getLocation()` further down in the call stack cause any behavioural change? As a start: would all tests still pass? (Tests do not always represent the precise behaviour, but they generally are pretty close to that.)
2. Can't we store the location capture in a continuation and make it available only when actually used?

Reply:
1. We can delegate the burden of computing the location to `LogEvent.getSource()` (actually most implementation already do it) and **document** it in the Javadoc. This way we can treat location the same way we treat `Message.getFormattedMessage()`: we call it just before a thread jump, otherwise we let the layouts do it. I prefer to do these changes in a separate PR.
2. The way it currently works the `location` parameter is passed all the way down to the `LogEvent`. Whether the location appears in the log or not follows the following logic:

- if `location` is not null (e.g. it was provided by the caller or Log4j Transform) it will be available,
- if `!isIncludeLocation() && requiresLocation()` it is computed,
- otherwise it will be unavailable.

Reply:
The actual implementation was a little bit different, see #2329.